### PR TITLE
Modifications for concourse CI

### PR DESCRIPTION
**Pull request by siredmar**
# Description
The default paths `/gen` and `/git` didn't match for working with Concourse CI. 
Per default the docker image behaves like before setting those default paths. However it is now possible to override these default paths with `GIT_PATH` and `GEN_PATH` environment variables.

The reason for this change is that concourse is not able to mount to those default directories. When starting concourse with input/output resources directories are created in `/tmp/build/<id>/` where you don't know the `<id>` (e.g. /tmp/build/2346634/source). The entrypoint script checks if either variable `GIT_PATH` and `GEN_PATH` is set and uses those as directories to work on.

### CODE REVIEW

File: gitversion/entrypoint.sh [hunk 1]
```diff
@@ -2,11 +2,46 @@
...
+  GEN=${p}
+else
+  # fallback
+  GEN="/gen"
+fi
 
 # check if /git is a git repository
-if [ ! -d "/git" ]; then
+if [ ! -d "${GIT}" ]; then
   echo "No git repository supplied. use docker -v PATH_TO_YOUR_REPO:/git"
```
Comment (mlilien) [inline gitversion/entrypoint.sh]: echo has to be adapted for ${GIT} vs /git

File: gitversion/inputfind [hunk 2]
```diff
@@ -0,0 +1,2 @@
+#!/bin/bash
+echo $(df | grep ${1} | awk '{print $6}')
```
Comment (mlilien) [inline gitversion/inputfind]: seems not to be a safe method.
on my device I get
```
echo $(df | grep /tmp | awk '{print $6}')
df: /home/vagrant/asp-device/out/itests/TestComponentStartStopImmediately_IntegrationTestBootup/target_opt: Transport endpoint is not connected
```
So what I understand here, you want to get the mountpoint of a dir. Can't you parse /etc/mtab directly?

File: gitversion/entrypoint.sh [hunk 3]
```diff
@@ -2,11 +2,46 @@
...
+    echo GEN_PATH is a command.
+    p=${d}
+  fi
+  # create path if not existent
+  [ ! -d "$p" ] &&  echo creating ${p} && mkdir -p ${p}
+  GEN=${p}
+else
+  # fallback
+  GEN="/gen"
+fi
```
Comment (mlilien) [inline gitversion/entrypoint.sh:15]: If I only read the entrypoint script ... I will not understand, why you did these changes. It feels cumbersome to adapt the container to concourse needs.

Quote: "This is needed for concourse as it handles input and output resources with a different approach. Those paths are not accessible from within the runner."
That means concouse doesn't allow own volume mounts? If so, isn't that a common problem with a common solution then?

**Review by mlilien — CHANGES_REQUESTED** [review]
I have problems to understand the reason why concourse needs an adapted container ... how do other projects solve this problem? The Problem is that you can not directly configure volume mounts, right?

File: gitversion/entrypoint.sh (hunk 1 above)
Comment (siredmar) [reply gitversion/entrypoint.sh]: You are right.

File: gitversion/inputfind (hunk 2 above)
Comment (siredmar) [reply gitversion/inputfind]: Did a refactoring with a much easier approach. I'll update the description of this PR to eloborate.

File: gitversion/entrypoint.sh (hunk 3 above)
Comment (siredmar) [reply gitversion/entrypoint.sh:15]: If you like you can see these changes more generic. 
The container now has the ability to not run on /git and /gen folders but got a generic way to set these working directories wherever you like. true, this is not needed when running locally with dobi where you just can mount to /gen and /git.0

However for the concourse case this does not work at all. You are fixed to the location /tmp/build/<id>/ (where you don't know which <id> will be used). 
So either you adapt the entrypoint script to make symlinks from /tmp/build/<id>/git to /git or create a way of generic setting the GIT_PATH.

Comment (mlilien) [reply gitversion/entrypoint.sh:15]: I can live with the technical solution for now. If we need more adaptions for usage with concourse in future, I would favour to overwrite the entrypoint  with an external one when using concourse.

**Review by mlilien — DISMISSED** [review]

File: gitversion/entrypoint.sh [hunk 4]
```diff
@@ -2,12 +2,21 @@
 set -o errexit   # abort on nonzero exitstatus
 set -o pipefail  # don't hide errors within pipes
 
-GIT=/git
-GEN=/gen
+if [ ! -z "$GIT_PATH" ]; then
```
Comment (FlorianReinhold) [inline gitversion/entrypoint.sh:5]: it would be great if the variable GIT_PATH and GEN_PATH were mentioned in the documentation.

**Review by FlorianReinhold — CHANGES_REQUESTED** [review]

File: gitversion/entrypoint.sh (hunk 4 above)
Comment (siredmar) [reply gitversion/entrypoint.sh:5]: Fixed.

**Review by FlorianReinhold — APPROVED** [review]

**Pull request merged**